go/apps/tao_launch/tao_launch.go

```diff
@@ -14,100 +14,696 @@
+		syscall.SIGHUP,  // tty hangup (e.g. via disown)
+		syscall.SIGABRT, // abort (e.g. via kill)
+		syscall.SIGUSR1, // user-defined (e.g. via kill)
+		syscall.SIGUSR2, // user-defined (e.g. via kill)
+	)
+
+	// Start the hosted program
+	subprin, pid, err := client.StartHostedProgram(wd, fds[:], binary, args...)
+	failIf(err, "Can't start hosted program")
+	fmt.Fprintf(noise, "[started hosted program with pid %d]\n", pid)
+	fmt.Fprintf(noise, "[subprin is %v]\n", subprin)
+
+	if pidOut != nil {
+		fmt.Fprintln(pidOut, pid)
+		pidOut.Close()
+	}
+
+	if nameOut != nil {
+		fmt.Fprintln(nameOut, subprin)
+		nameOut.Close()
+	}
+
+	if disown || daemon {
+		return
+	}
+
+	// Listen for exit status from host
+	status := make(chan int, 1)
+	go func() {
+		s, _ := client.WaitHostedProgram(pid, subprin)
+		// For short programs, we often lose the race, so
+		// we get a "no such hosted program" error.
+		// failIf(err, "Can't wait for hosted program exit")
+		status <- s
+	}()
+
+	wasForeground := false
+	if proxying && isForeground() {
+		fmt.Fprintf(noise, "[in foreground]\n")
+		dropSIGTTIN()
+		wasForeground = true
+	}
+
+	// Proxy stdin, if needed
+	if proxying {
+		pr.Close()
+		go func() {
+			_, err := io.Copy(pw, os.Stdin)
+			warnIf(err, "Can't copy from stdin to pipe")
+			pw.Close()
+		}()
+	}
+
+	// If we are proxying and (were) background, we should probably
+	// have done a read() by now and gotten SIGTTIN and stopped. Let's
+	// pause a moment to be sure the read() happens.
+	time.Sleep(moment)
+
+	// By this point, if we had been foreground, we might still be. Or, we might
+	// have been foreground but just gotten SIGTSTP and are now madly dropping
+	// SIGTTIN until we get into the loop below to handle the SIGTSTP.
+	//
+	// Alternatively, if we had been background, we would have been stopped by
+	// the default SIGTTIN, so the only way we would be here is if we later got
+	// pulled foreground via fg. We want to be dropping SIGTTIN in case we get a
+	// SIGTSTP.
+	if proxying && !wasForeground {
+		dropSIGTTIN()
+	}
+
+	next := cont
+	for next != done {
+		select {
+		case s := <-status:
+			fmt.Fprintf(noise, "[hosted program exited, %s]\n", exitCode(s))
+			next = done
+		case sig := <-signals:
+			next = handle(sig, pid)
+		}
+		if next == resumed && proxying && !isForeground() {
+			// Need to toggle SIGTTIN handling and block (that's the only way to
+			// stop spinning on SIGTTIN), but only after handling all pending
+			// signals (e.g. SIGCONT then SIGHUP, or SIGCONT then SIGTERM).
+			for next == resumed {
+				select {
+				case s := <-status:
+					fmt.Fprintf(noise, "[hosted program exited, %s]\n", exitCode(s))
+					next = done
+				case sig := <-signals:
+					next = handle(sig, pid)
+					if next == cont {
+						next = resumed
+					}
+				default:
+					next = cont
+					defaultSIGTTIN()
+					time.Sleep(moment)
+					dropSIGTTIN()
+				}
+			}
+		}
+	}
+	signal.Stop(signals)
+}
+
+type todo int
```

Comment:
Because it encodes the next action the main loop is meant "to do". I agree it looks weird. Suggestions? Use int? This whole loop is a mess, but believe me I tried hard to make this comprehensible.